src/genesis/observers/balances.py

```diff
@@ -64,13 +68,60 @@ def __init__(self, db_conn: Connection, on_completed=None, on_error=None) -> Non
                                                 on_completed=on_completed,
                                                 on_error=on_error)
 
+    def _get_name_and_index(self, e: UniqueViolation, balances: List[Balance]) -> Tuple[str, Tuple[int, int]]:
+        # Extract account name and coin from error string
+        duplicate_balance = str(e).split("(")[2].split(")")[0]
+
+        # Find duplicate account id
+        duplicate_balance_index = None
+        duplicate_coin_index = None
+        for i in range(len(balances)):
+            for j in range(len(balances[i].coins)):
+                if f"{balances[i].address}-{balances[i].coins[j].denom}" in duplicate_balance:
+                    duplicate_balance_index = i
+                    duplicate_coin_index = j
+
+        return duplicate_balance, duplicate_balance_index, duplicate_coin_index
+
     def copy_balances(self, balances: List[Balance]) -> None:
         with self._db_conn.cursor() as db:
-            with db.copy(f'COPY {self._table} ({",".join(self.column_names)}) FROM STDIN') as copy:
-                for balance in balances:
-                    for coin in balance.coins:
-                        id_ = f"{balance.address}-{coin.denom}"
-                        copy.write_row((f"{v}" for v in (id_, balance.address, coin.amount, coin.denom)))
+            duplicate_occured = True
+
+            while duplicate_occured:
+                try:
+                    duplicate_occured = False
+                    with db.copy(f'COPY {self._table} ({",".join(self.column_names)}) FROM STDIN') as copy:
+                        for balance in balances:
+                            for coin in balance.coins:
+                                id_ = f"{balance.address}-{coin.denom}"
```

Comment:
nit: Perhaps this ID derivation is worth factoring out to a function as we're doing it in a couple of places now.

Reply:
There are 2 similar but different usages. It adds together address and denom.

I think function like this would be almost pointless.
`
    @staticmethod
    def _addr_and_denom_to_id(address: str, denom: str):
        return f"{address}-{denom}"
`